src/module.c

```diff
@@ -12388,28 +12388,30 @@ int moduleLoad(const char *path, void **module_argv, int module_argc, int is_loa
 /* Unload the module registered with the specified name. On success
  * C_OK is returned, otherwise C_ERR is returned and errmsg is set
  * with an appropriate message. */
-int moduleUnload(sds name, const char **errmsg) {
+int moduleUnload(sds name, const char **errmsg, int forced_unload) {
     struct RedisModule *module = dictFetchValue(modules,name);
 
-    if (module == NULL) {
-        *errmsg = "no such module with that name";
-        return C_ERR;
-    } else if (listLength(module->types)) {
-        *errmsg = "the module exports one or more module-side data "
-                  "types, can't unload";
-        return C_ERR;
-    } else if (listLength(module->usedby)) {
-        *errmsg = "the module exports APIs used by other modules. "
-                  "Please unload them first and try again";
-        return C_ERR;
-    } else if (module->blocked_clients) {
-        *errmsg = "the module has blocked clients. "
-                  "Please wait for them to be unblocked and try again";
-        return C_ERR;
-    } else if (moduleHoldsTimer(module)) {
-        *errmsg = "the module holds timer that is not fired. "
-                  "Please stop the timer or wait until it fires.";
-        return C_ERR;
+    if (!forced_unload) {
```

Comment:
are we sure about all of these?
i can understand why we may want to ignore registered data types if the loading failed and proceed with unload (there should not yet be any keys of these data types).

but the part about blocked clients (probably unreachable), and timer events (could be reachable), doesn't seem right.

instead of all this indentation, i'd add the `!force` condition to each of the relevant checks.

Reply:
@oranagra We are only sure about the case with registered data types, and we only tested that case, so I added the `!force` condition only to it. Thanks.